pkg/image/cdi/cdi.go

```diff
@@ -1,32 +1,269 @@
 package cdi
 
 import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+
+	ctlcorev1 "github.com/rancher/wrangler/v3/pkg/generated/controllers/core/v1"
+	"github.com/sirupsen/logrus"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	cdiv1 "kubevirt.io/containerized-data-importer-api/pkg/apis/core/v1beta1"
+
 	harvesterv1 "github.com/harvester/harvester/pkg/apis/harvesterhci.io/v1beta1"
+	ctlcdiv1 "github.com/harvester/harvester/pkg/generated/controllers/cdi.kubevirt.io/v1beta1"
 	"github.com/harvester/harvester/pkg/image/backend"
+	"github.com/harvester/harvester/pkg/image/common"
 )
 
 type Backend struct {
+	ctx              context.Context
+	dataVolumeClient ctlcdiv1.DataVolumeClient
+	pvcCache         ctlcorev1.PersistentVolumeClaimCache
+	vmio             common.VMIOperator
+
+	dataVolumeCreated bool
 }
 
-func GetBackend() backend.Backend {
-	return &Backend{}
+func GetBackend(ctx context.Context, dataVolumeClient ctlcdiv1.DataVolumeClient, pvcCache ctlcorev1.PersistentVolumeClaimCache, vmio common.VMIOperator) backend.Backend {
+	return &Backend{
+		ctx:               ctx,
+		dataVolumeClient:  dataVolumeClient,
+		pvcCache:          pvcCache,
+		vmio:              vmio,
+		dataVolumeCreated: false,
+	}
 }
 
-func (cb *Backend) Initialize(vmi *harvesterv1.VirtualMachineImage) (*harvesterv1.VirtualMachineImage, error) {
-	return vmi, nil
+func (b *Backend) Initialize(vmImg *harvesterv1.VirtualMachineImage) (*harvesterv1.VirtualMachineImage, error) {
+	// if dataVolume is already created, return
+	created, err := b.isDataVolumeCreated(vmImg)
+	if err != nil {
+		return vmImg, fmt.Errorf("failed to check DataVolume: %v", err)
+	}
+	if created {
+		return vmImg, nil
+	}
+
+	switch vmImg.Spec.SourceType {
+	case harvesterv1.VirtualMachineImageSourceTypeDownload:
+		return b.initializeDownload(vmImg)
+	case harvesterv1.VirtualMachineImageSourceTypeUpload:
+		// do nothing when vmimage source is upload, dataVolume will be created by upload handler
+		return vmImg, nil
+	case harvesterv1.VirtualMachineImageSourceTypeExportVolume:
+		return b.initializeExportFromVolume(vmImg)
+	default:
+		return vmImg, fmt.Errorf("unsupported source type: %s", vmImg.Spec.SourceType)
+	}
 }
 
-func (cb *Backend) Check(_ *harvesterv1.VirtualMachineImage) error {
+func (b *Backend) Check(vmImg *harvesterv1.VirtualMachineImage) error {
+	targetDVNs := vmImg.ObjectMeta.Namespace
+	targetDVName := vmImg.ObjectMeta.Name
+	targetDV, err := b.dataVolumeClient.Get(targetDVNs, targetDVName, metav1.GetOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		return fmt.Errorf("failed to get DataVolume %s/%s: %v", targetDVNs, targetDVName, err)
+	}
+	if apierrors.IsNotFound(err) {
+		logrus.Info("DataVolume not found, waiting for the initialization")
+		return err
+	}
```

Comment:
Maybe can be compacted?
```suggestion
	if err != nil {
		if apierrors.IsNotFound(err) {
			logrus.Info("DataVolume not found, waiting for the initialization")
		}
		return fmt.Errorf("failed to get DataVolume %s/%s: %v", targetDVNs, targetDVName, err)
	}
```